File: src/com/reco1l/framework/net/Requester.java
package com.reco1l.framework.net;

import androidx.core.util.Consumer;
import com.reco1l.framework.util.Logging;
import okhttp3.*;

import java.util.Objects;
import java.util.concurrent.TimeUnit;

/**
 * @author Reco1l
 */
public class Requester implements AutoCloseable
{

    public static final int DEFAULT_TIMEOUT = 10000;

    public static final OkHttpClient DEFAULT_CLIENT = new OkHttpClient
            .Builder()
            .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS)
            .build();

    //----------------------------------------------------------------------------------------------------------------//

    protected OkHttpClient mClient;
    protected Request mRequest;

    protected ResponseBody mResponseBody;
    protected Response mResponse;
    protected Call mCall;

    protected QueryContent mQuery;

    //----------------------------------------------------------------------------------------------------------------//

    public Requester()
    {
        this(null);
    }

    public Requester(String url)
    {
        super();
        mClient = DEFAULT_CLIENT;

        if (url != null)
        {
            mRequest = new Request
                    .Builder()
                    .url(url)
                    .build();
        }
    }

    //----------------------------------------------------------------------------------------------------------------//

    public void setClient(OkHttpClient client)
    {
        mClient = client;
    }

    public void setRequest(Request request)
    {
        mRequest = request;
    }

    public void setQuery(QueryContent query)
    {
        mQuery = query;
    }

    //----------------------------------------------------------------------------------------------------------------//

    public void buildClient(Consumer<OkHttpClient.Builder> consumer)
    {
        var builder = mClient != null ? mClient.newBuilder() : new OkHttpClient.Builder();

        if (consumer != null)
        {
            consumer.accept(builder);
        }
        mClient = builder.build();
    }

    public void buildRequest(Consumer<Request.Builder> consumer)
    {
        var builder = mRequest != null ? mRequest.newBuilder() : new Request.Builder();

        if (consumer != null)
        {
            consumer.accept(builder);
        }
        mRequest = builder.build();
    }

    //----------------------------------------------------------------------------------------------------------------//

    public OkHttpClient getClient()
    {
        return mClient;
    }

    public Request getRequest()
    {
        return mRequest;
    }

    //----------------------------------------------------------------------------------------------------------------//

    public Call getCall()
    {
        Objects.requireNonNull(mCall, "You need to call execute() first!");
        return mCall;
    }

    public Response getResponse()
    {
        Objects.requireNonNull(mResponse, "You need to call execute() first!");
        return mResponse;
    }

    public ResponseBody getResponseBody()
    {
        Objects.requireNonNull(mResponseBody, "You need to call execute() first!");
        return mResponseBody;
    }

    //----------------------------------------------------------------------------------------------------------------//

    /**
     * Called as soon the response and response body is created.
     */
    protected void onResponse(Response response) throws Exception
    {
    }

    /**
     * Called when response code isn't 200-299, this can also be called from inherited onResponseSuccess() exception thrown.
     */
    protected void onResponseFail(Exception exception) throws Exception
    {
    }

    /**
     * Called when response code is 200-299.
     */
    protected void onResponseSuccess(Response response) throws Exception
    {
    }

    //----------------------------------------------------------------------------------------------------------------//

    @Override
    public void close()
    {
        if (mResponse != null)
        {
            mResponse.close();

            if (mResponseBody != null)
            {
                mResponseBody.close();
            }
        }
    }

    //----------------------------------------------------------------------------------------------------------------//

    /**
     * Make sure to call before call close() or inside the try-with-resources statement.
     */
    public Requester execute() throws Exception
    {
        Objects.requireNonNull(mClient, "Client cannot be null!");
        Objects.requireNonNull(mRequest, "Request cannot be null!");

        if (mQuery != null)
        {
            mRequest = mRequest.newBuilder()
                    .url(mRequest.url() + mQuery.toString())
                    .build();
        }

        Logging.i(this, "Inserted URL: " + mRequest.url());

        mCall = mClient.newCall(mRequest);
        try
        {
            mResponse = mCall.execute();
            mResponseBody = mResponse.body();

            onResponse(mResponse);

            if (mResponse.isSuccessful())
            {
                onResponseSuccess(mResponse);
            }
            else
            {
                throw new ResponseException(mResponse);
            }
        }
        catch (Exception e)
        {
            onResponseFail(e);
            throw e;
        }
        return this;
    }

    /**
     * Make sure to call before call close() or inside the try-with-resources statement.
     *
     * @return The response body.
     */
    public ResponseBody executeAndGetBody() throws Exception
    {
        //noinspection resource
        return execute().getResponseBody();
    }
}

File: src/com/reco1l/framework/net/Requester.kt
package com.reco1l.framework.net

import com.reco1l.framework.extensions.className
import com.reco1l.framework.extensions.logIfDebug
import okhttp3.*
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * @author Reco1l
 */
open class Requester(url: String) : AutoCloseable
{

    /**
     * Indicates if the requester should log responses.
     */
    var log = true

    /**
     * The request query.
     */
    var query: QueryContent? = null

    /**
     * The call executed by the client, don't use it before calling [execute].
     */
    lateinit var call: Call

    /**
     * The response get, don't use it before calling [execute].
     */
    lateinit var response: Response

    /**
     * The response body, don't use it before calling [execute].
     */
    lateinit var responseBody: ResponseBody



    private var client: OkHttpClient? = DEFAULT_CLIENT

    private var request: Request? = defaultRequest(url)



    fun buildClient(unit: (OkHttpClient.Builder) -> Unit)
    {
        val builder = client?.newBuilder() ?: OkHttpClient.Builder()

        unit(builder)
        client = builder.build()
    }

    fun buildRequest(unit: (Request.Builder) -> Unit)
    {
        val builder = request?.newBuilder() ?: Request.Builder()

        unit(builder)
        request = builder.build()
    }



    /**
     * Called as soon the response and response body is created.
     */
    @Throws(Exception::class)
    protected open fun onResponse(response: Response?)
    {
    }

    /**
     * Called when response code isn't 200-299, this can also be called from inherited [onResponseSuccess] exception thrown.
     */
    @Throws(Exception::class)
    protected fun onResponseFail(exception: Exception)
    {
        throw exception
    }

    /**
     * Called when response code is 200-299.
     */
    @Throws(Exception::class)
    protected open fun onResponseSuccess(response: Response?)
    {
    }



    override fun close()
    {
        if (::response.isInitialized)
        {
            response.close()
        }
        if (::responseBody.isInitialized)
        {
            responseBody.close()
        }
    }



    /**
     * Make sure to call before call close() or inside the try-with-resources statement.
     */
    @Throws(Exception::class)
    open fun execute(): Requester
    {
        requireNotNull(client)
        requireNotNull(request)

        if (query != null)
        {
            request = request!!.newBuilder().url("${request!!.url}$query").build()
        }

        call = client!!.newCall(request!!)

        try
        {
            if (log)
                "Inserted url: ${request!!.url}".logIfDebug(className)

            response = call.execute()
            responseBody = response.body!!

            onResponse(response)

            if (response.isSuccessful)
            {
                onResponseSuccess(response)
            }
            else
            {
                throw ResponseException(response)
            }
        }
        catch (e: Exception)
        {
            onResponseFail(e)
        }
        return this
    }

    /**
     * Make sure to call before call [close] or inside the try-with-resources statement.
     *
     * @return The response body.
     */
    @Throws(Exception::class)
    fun executeAndGetBody(): ResponseBody = execute().responseBody



    companion object
    {
        const val DEFAULT_TIMEOUT = 10000

        val DEFAULT_CLIENT = OkHttpClient
                .Builder()
                .connectTimeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                .build()

        fun defaultRequest(url: String) = Request
                .Builder()
                .url(url)
                .build()
    }
}
